File: packages/server/src/services/depeg.service.ts
import { DepegAlert } from '@shared';
import { CoinGeckoStablecoin, CoinGeckoCoinPlatforms } from '../api/coingecko.stablecoins.api.service';

/** Assumed peg for all USD stablecoins */
const USD_PEG = 1.0;

/** Deviations above these thresholds trigger alerts */
const WARN_THRESHOLD = 0.005; // 0.5%
const CRITICAL_THRESHOLD = 0.02; // 2%

/**
 * Builds a normalised symbol → current price map from CoinGecko stablecoin data.
 * Symbols are lowercased for case-insensitive matching.
 */
export function buildStablecoinPriceMap(coins: CoinGeckoStablecoin[]): Map<string, number> {
  const map = new Map<string, number>();
  for (const coin of coins) {
    if (coin.current_price != null) {
      map.set(coin.symbol.toLowerCase(), coin.current_price);
    }
  }
  return map;
}

/**
 * Builds a contract address → stablecoin symbol map by cross-referencing:
 *  - stablecoins: the price data we already have (id → symbol)
 *  - coinList:    the full coin list with on-chain addresses (id → platforms → address)
 *
 * Addresses and symbols are lowercased. Only stablecoin IDs present in the price
 * data are indexed, so the symbol can then be used to look up the price in priceMap.
 */
export function buildStablecoinAddressMap(
  stablecoins: CoinGeckoStablecoin[],
  coinList: CoinGeckoCoinPlatforms[],
): Map<string, string> {
  const idToSymbol = new Map<string, string>();
  for (const coin of stablecoins) {
    if (coin.current_price != null) {
      idToSymbol.set(coin.id, coin.symbol.toLowerCase());
    }
  }

  const addressMap = new Map<string, string>();
  for (const coin of coinList) {
    const symbol = idToSymbol.get(coin.id);
    if (symbol == null) continue;
    for (const address of Object.values(coin.platforms)) {
      if (address && address.startsWith('0x')) {
        addressMap.set(address.toLowerCase(), symbol);
      }
    }
  }
  return addressMap;
}

/**
 * Parses a pool symbol string into individual token symbols.
 * Handles separators: -, /, +, space.
 * e.g. "USR-USDC" → ["USR", "USDC"], "crvUSD" → ["crvUSD"]
 */
export function parsePoolSymbols(poolSymbol: string): string[] {
  return poolSymbol
    .split(/[-/+\s]+/)
    .map((s) => s.trim())
    .filter(Boolean);
}

function makeAlert(symbol: string, price: number): DepegAlert | null {
  const deviation = (price - USD_PEG) / USD_PEG;
  const absDeviation = Math.abs(deviation);
  if (absDeviation < WARN_THRESHOLD) return null;
  return {
    symbol,
    currentPrice: price,
    pegPrice: USD_PEG,
    deviation,
    severity: absDeviation >= CRITICAL_THRESHOLD ? 'critical' : 'warning',
  };
}

/**
 * Checks a pool for stablecoin depeg risk using two sources:
 *  1. Pool symbol tokens looked up in the symbol → price map
 *  2. underlyingTokens contract addresses looked up in the address → symbol map,
 *     then prices resolved via the symbol → price map
 *
 * Applies to all pools — the price maps act as the natural filter
 * (only known stablecoins produce alerts).
 */
export function checkDepeg(
  poolSymbol: string,
  priceMap: Map<string, number>,
  underlyingTokens: string[] | null = null,
  addressMap: Map<string, string> | null = null,
): DepegAlert[] {
  const alerts: DepegAlert[] = [];
  const seen = new Set<string>(); // deduplicate by lowercase symbol key

  // --- Symbol-based check ---
  for (const token of parsePoolSymbols(poolSymbol)) {
    const key = token.toLowerCase();
    if (seen.has(key)) continue;
    seen.add(key);
    const price = priceMap.get(key);
    if (price == null) continue;
    const alert = makeAlert(token, price);
    if (alert) alerts.push(alert);
  }

  // --- Address-based check (underlyingTokens) ---
  if (addressMap && underlyingTokens) {
    for (const address of underlyingTokens) {
      const symbol = addressMap.get(address.toLowerCase());
      if (symbol == null) continue; // address is not a known stablecoin
      if (seen.has(symbol)) continue; // already reported via symbol check
      seen.add(symbol);
      const price = priceMap.get(symbol);
      if (price == null) continue;
      const alert = makeAlert(symbol.toUpperCase(), price);
      if (alert) alerts.push(alert);
    }
  }

  return alerts;
}

File: src/Credfeto.Defi.Server/Services/DepegService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;
using Credfeto.Defi.Server.Models;

namespace Credfeto.Defi.Server.Services;

/// <summary>
///     Builds stablecoin price maps and detects depeg risk for pool tokens.
/// </summary>
public static partial class DepegService
{
    private const decimal USD_PEG = 1.0m;
    private const decimal WARN_THRESHOLD = 0.005m; // 0.5%
    private const decimal CRITICAL_THRESHOLD = 0.02m; // 2%

    [GeneratedRegex(pattern: "[-/+\\s]+", options: RegexOptions.ExplicitCapture, matchTimeoutMilliseconds: 500)]
    private static partial Regex TokenSplitRegex();

    /// <summary>
    ///     Builds a normalised symbol → current price map from CoinGecko stablecoin data.
    ///     Symbols are lowercased for case-insensitive matching.
    /// </summary>
    public static IReadOnlyDictionary<string, decimal> BuildStablecoinPriceMap(IReadOnlyList<CoinGeckoStablecoin> coins)
    {
        Dictionary<string, decimal> map = new(StringComparer.OrdinalIgnoreCase);

        foreach (CoinGeckoStablecoin coin in coins)
        {
            if (coin.CurrentPrice.HasValue)
            {
                map.TryAdd(key: coin.Symbol.ToLowerInvariant(), value: coin.CurrentPrice.Value);
            }
        }

        return map;
    }

    /// <summary>
    ///     Builds a contract address → stablecoin symbol map from the full CoinGecko coin list.
    ///     Only stablecoin IDs present in the price data are indexed.
    /// </summary>
    public static IReadOnlyDictionary<string, string> BuildStablecoinAddressMap(
        IReadOnlyList<CoinGeckoStablecoin> stablecoins,
        IReadOnlyList<CoinGeckoCoinPlatforms> coinList
    )
    {
        Dictionary<string, string> idToSymbol = new(StringComparer.OrdinalIgnoreCase);

        foreach (CoinGeckoStablecoin coin in stablecoins)
        {
            if (coin.CurrentPrice.HasValue)
            {
                idToSymbol.TryAdd(key: coin.Id, value: coin.Symbol.ToLowerInvariant());
            }
        }

        Dictionary<string, string> addressMap = new(StringComparer.OrdinalIgnoreCase);

        foreach (CoinGeckoCoinPlatforms coin in coinList)
        {
            if (!idToSymbol.TryGetValue(key: coin.Id, out string? symbol))
            {
                continue;
            }

            if (coin.Platforms is null)
            {
                continue;
            }

            foreach (
                string address in coin.Platforms.Values.Where(a =>
                    !string.IsNullOrEmpty(a)
                    && a.StartsWith(value: "0x", comparisonType: StringComparison.OrdinalIgnoreCase)
                )
            )
            {
                addressMap.TryAdd(key: address.ToLowerInvariant(), value: symbol);
            }
        }

        return addressMap;
    }

    /// <summary>
    ///     Parses a pool symbol string into individual token symbols.
    ///     Handles separators: -, /, +, space.
    ///     e.g. "USR-USDC" → ["USR", "USDC"], "crvUSD" → ["crvUSD"]
    /// </summary>
    public static string[] ParsePoolSymbols(string poolSymbol)
    {
        return
        [
            .. TokenSplitRegex()
                .Split(poolSymbol)
                .Select(static p => p.Trim())
                .Where(static p => !string.IsNullOrEmpty(p)),
        ];
    }

    /// <summary>
    ///     Checks a pool for stablecoin depeg risk using two sources:
    ///     1. Pool symbol tokens looked up in the symbol → price map.
    ///     2. Underlying token contract addresses looked up in the address → symbol map,
    ///        then prices resolved via the symbol → price map.
    /// </summary>
    /// <summary>
    ///     Overload accepting a <see cref="ReadOnlySpan{T}" /> of underlying token addresses.
    ///     Delegates to the array overload after converting.
    /// </summary>
    public static IReadOnlyList<DepegAlert> CheckDepeg(
        string poolSymbol,
        IReadOnlyDictionary<string, decimal> priceMap,
        ReadOnlySpan<string> underlyingTokens,
        IReadOnlyDictionary<string, string>? addressMap
    )
    {
        return CheckDepeg(
            poolSymbol: poolSymbol,
            priceMap: priceMap,
            underlyingTokens: underlyingTokens.IsEmpty ? null : underlyingTokens.ToArray(),
            addressMap: addressMap
        );
    }

    /// <summary>
    ///     Checks a pool for stablecoin depeg risk using two sources:
    ///     1. Pool symbol tokens looked up in the symbol → price map.
    ///     2. Underlying token contract addresses looked up in the address → symbol map,
    ///        then prices resolved via the symbol → price map.
    /// </summary>
    public static IReadOnlyList<DepegAlert> CheckDepeg(
        string poolSymbol,
        IReadOnlyDictionary<string, decimal> priceMap,
        string[]? underlyingTokens,
        IReadOnlyDictionary<string, string>? addressMap
    )
    {
        List<DepegAlert> alerts = [];
        HashSet<string> seen = new(StringComparer.OrdinalIgnoreCase);

        // Symbol-based check
        foreach (string token in ParsePoolSymbols(poolSymbol))
        {
            string key = token.ToLowerInvariant();

            if (!seen.Add(key))
            {
                continue;
            }

            if (!priceMap.TryGetValue(key: key, out decimal price))
            {
                continue;
            }

            DepegAlert? alert = MakeAlert(symbol: token, price: price);

            if (alert is not null)
            {
                alerts.Add(alert);
            }
        }

        // Address-based check (underlyingTokens)
        if (addressMap is not null && underlyingTokens is not null)
        {
            foreach (string address in underlyingTokens)
            {
                if (!addressMap.TryGetValue(key: address.ToLowerInvariant(), out string? symbol))
                {
                    continue;
                }

                if (!seen.Add(symbol))
                {
                    continue;
                }

                if (!priceMap.TryGetValue(key: symbol, out decimal price))
                {
                    continue;
                }

                DepegAlert? alert = MakeAlert(symbol: symbol.ToUpperInvariant(), price: price);

                if (alert is not null)
                {
                    alerts.Add(alert);
                }
            }
        }

        return alerts;
    }

    private static DepegAlert? MakeAlert(string symbol, decimal price)
    {
        decimal deviation = (price - USD_PEG) / USD_PEG;
        decimal absDeviation = Math.Abs(deviation);

        if (absDeviation < WARN_THRESHOLD)
        {
            return null;
        }

        return new DepegAlert
        {
            Symbol = symbol,
            CurrentPrice = price,
            PegPrice = USD_PEG,
            Deviation = deviation,
            Severity = absDeviation >= CRITICAL_THRESHOLD ? "critical" : "warning",
        };
    }
}
